### PR TITLE
Add developer productivity scripts and add testing guidelines

### LINKED ISSUE #852: Add developer productivity scripts and document testing guidelines
**How to categorize this issue?**
/area dev-productivity
/area testing
/kind enhancement

**What would you like to be added**:
Introduce the following capabilities:

- [ ] Add capability to stress test to help de-flake tests (script and/or make target) - for both unit and integration-tests
- [ ] Add capability to start envtest separately and run integration-tests targeting it. This helps analysing the state of the resources and helps in debugging tests.
- [ ] Add capability to check if there are incompatible api changes.
- [ ] Add a common formatter (make target) for all source files to ensure uniformity on how all code needs to be formatted.
- [ ] Add a doc on testing guidelines

**Why is this needed**:
Enhance developer productivity

### PR DESCRIPTION
**Pull request by unmarshall**
**How to categorize this PR?**

/area dev-productivity
/area documentation
/kind test
/kind enhancement

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #852 

**Special notes for your reviewer**:

**Release note**:

```other developer
Introduced testing guidelines, added developer productivity scripts and make targets to stress test, debug integration tests, formatting and detecting incompatible api changes.
```

### CODE REVIEW

File: docs/development/testing.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+* In CI pipelines e2e tests are run with S3 compatible [LocalStack](https://www.localstack.cloud/) (in cases where backup functionality has been enabled for an etcd cluster).
+
+  > In future we will only be using a file-system based local provider to reduce the run times for the e2e tests when run in a CI pipeline.
+
+* e2e tests can be triggered either with other cloud provider object-store emulators or they can also be run against actual/remove cloud provider object-store services.
+
+* In contrast to integration tests, in e2e tests, it might make sense to specify higher timeouts for Gomega's [AsyncAssertion](https://pkg.go.dev/github.com/onsi/gomega#AsyncAssertion) calls.
+
+### Running e2e Tests
```
Comment (ashwani2k) [inline docs/development/testing.md:164]:
```suggestion
### Running e2e tests locally
```

**Review by ashwani2k — CHANGES_REQUESTED** [review]
Thanks for this well written document. 
I'm assuming once merged it it get auto updated to the persona docs https://gardener.cloud/docs/other-components/etcd-druid/local-e2e-tests/. 

Also you might want to relook at how this is getting rendered in the webiste. 
We have -- 
https://gardener.cloud/docs/other-components/etcd-druid/readme/ 
and 
https://gardener.cloud/docs/other-components/etcd-druid/ 
If I click this I can't see any way to reach to concepts and proposals in the website.

File: hack/check-apidiff.sh [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply hack/check-apidiff.sh:1]: check-apidiff.sh in gardener is coded to look at specific set of API packages which does not cover etcd and it ideally should not as well. Each component which provides its own API should do this check independently. At this time i do not see any need to give reference to the original source. 

>Also I'm not sure if its necessary to deviate in this script apart from exported_pkg list in Line 24

Improvements for readability should always be made to the script. This is what i have done. Intent is never to be in sync with the source.

File: hack/format.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply hack/format.sh:1]: There is no reason to use 2 tools and i have questioned that. I have achieve the formatting with just using one tool and that for us is quite sufficient.

File: hack/test-integration.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply hack/test-integration.sh:1]: Ah yes, in the `Makefile` i am currently not using this script. So in this PR i will perhaps remove this script and work on it later.

File: docs/development/testing.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+Run all unit tests
+
+```bash
+> make test-unit
+```
+
+Run unit tests of specific packages
+
+```bash
+> ./hack/test-go.sh <package-1> <package-2>
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply docs/development/testing.md:64]: Ah ok, the issue is that we install the tool `gotestfmt` as part of make target dependency. So when you run `make test-unit` then it will work as it installs the tool. I will add that instruction and then it should run for you. Good catch! I did not catch that as i already had this tool installed and in my path.

File: docs/development/testing.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+An example invocation:
+
+```bash
+> ./hack/stress-test.sh test-package=./internal/utils test-func=TestRunConcurrentlyWithAllSuccessfulTasks tool-params="-p 10"
+5s: 877 runs so far, 0 failures
+10s: 1906 runs so far, 0 failures
+15s: 2885 runs so far, 0 failures
+...
+```
+`stress` tool will output a path to a file containing the full failure message when a test run fails.
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply docs/development/testing.md:87]: You mean introduce flakiness and then run this test to figure that out? If yes, is that really required?

File: docs/development/testing.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+* We use gomega as our matcher or assertion library. Refer to Gomega's [official documentation](https://onsi.github.io/gomega/) for details regarding its installation and application in tests.
+
+* For naming the individual test/helper functions, ensure that the name describes what the function tests/helps-with. Naming is important for code readability even when writing tests.
+
+* Introduce helper functions for assertions to make test more readable where applicable - [example-assertion-function](https://github.com/gardener/etcd-druid/blob/4e9971aba3c3880a4cb6583d05843eabb8ca1409/test/it/controller/etcd/assertions.go#L117).
+
+* Introduce custom matchers to make tests more readable where applicable - [example-custom-matcher](https://github.com/gardener/etcd-druid/blob/4e9971aba3c3880a4cb6583d05843eabb8ca1409/test/utils/matcher.go#L89).
+
+* Do not use `time.Sleep` and friends as it renders the tests flaky.
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply docs/development/testing.md:25]: I mean you should not use sleep or its alternatives. Also these are common guidelines. If you see integration test section then i have mentioned using `gomega`'s AsyncAssertions. But this cannot be put into a common guideline as they do not apply for unit tests in general.

File: docs/development/testing.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+* If there are a lot of edge cases, extract dedicated functions that cover them and use unit tests to test them.
+
+### Running Unit Tests
+
+> **NOTE:** For unit tests we are currently transitioning away from [ginkgo](https://github.com/onsi/ginkgo) to using golang native tests. The `make test-unit` target runs both ginkgo and golang native tests. Once the transition is complete this target will be simplified.
+
+Run all unit tests
+
+```bash
+> make test-unit
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply docs/development/testing.md:58]: This is not possible right now. It is because we are using a tool to beautify the output which otherwise is harder to read. In the current output format the coverage is very minimally included as % against a package. So even if i introduce 2 variants one with coverage and one without the difference in verbosity will not be much, at least not enough to trouble you.

I will be replacing this tool with something that i will develop separately. I do not like any tool that is currently available in the community. Ginkgo uses a formatter which they have internally developed. We currently cannot use this as it is not an independent component.

File: docs/development/testing.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (unmarshall) [reply docs/development/testing.md:64]: Done

File: hack/test-integration.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (unmarshall) [reply hack/test-integration.sh]: removed the script as it is not used in the `Makefile` currently. Will change a few scripts in a later PR.

File: docs/development/testing.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+For any new contributions **tests are a strict requirement**. `Boy Scouts Rule` is followed: If you touch a code for which either no tests exist or coverage is insufficient then it is expected that you will add relevant tests.
+
+## Common guidelines for writing tests
+
+* We use the `Testing` package provided by the standard library in golang for writing all our tests. Refer to its [official documentation](https://pkg.go.dev/testing) to learn how to write tests using `Testing` package. You can also refer to [this](https://go.dev/doc/tutorial/add-a-test) example.
+
+* We use gomega as our matcher or assertion library. Refer to Gomega's [official documentation](https://onsi.github.io/gomega/) for details regarding its installation and application in tests.
+
+* For naming the individual test/helper functions, ensure that the name describes what the function tests/helps-with. Naming is important for code readability even when writing tests.
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply docs/development/testing.md]: Provided.

File: Makefile [hunk 10]
```diff
@@ -51,36 +66,26 @@ update-dependencies:
...
-	@go generate "$(REPO_ROOT)/internal/..."
-	@"$(HACK_DIR)/update-codegen.sh"
 
+# Rules for testing (unit, integration and end-2-end)
+# -------------------------------------------------------------------------
 # Run tests
-.PHONY: test
-test: $(GINKGO) $(GOTESTFMT)
+.PHONY: test-unit
+test-unit: $(GINKGO) $(GOTESTFMT)
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply Makefile:88]: I am confused, isn't it already `test-unit`?

[the remark this answers is not in the document]
Comment (unmarshall) [reply Makefile:88]: made the change.

File: docs/development/testing.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ashwani2k) [reply docs/development/testing.md:90]: Not really but I run this right now, it runs for ever and I was wondering what should one look out for if its flaky. So If I understand run this tool and see if your tests consistently fail and if so then its not a flaky test. Is that the intent?

File: hack/prepare-envtest.sh [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+ENVTEST_K8S_VERSION=${ENVTEST_K8S_VERSION:-"1.30"}
+
+echo "> Installing envtest tools@${ENVTEST_K8S_VERSION} with setup-envtest if necessary"
+if ! command -v setup-envtest &> /dev/null ; then
+  >&2 cat <<EOF
+  setup-envtest not available
+  You can either directly install it by following the instructions at https://pkg.go.dev/sigs.k8s.io/controller-runtime/tools/setup-envtest#section-readme
+  Or you can use `make test-integration` which will implicitly install it for you.
+EOF
```
Comment (seshachalam-yv) [inline hack/prepare-envtest.sh:19]: When the `setup-envtest` command is not available, the script mistakenly proceeds to execute the block below, leading to `make test-integration` being unintentionally triggered. 


To fix this, I recommend escaping the backticks for `make test-integration` in the output message, so it doesn't inadvertently get executed.


```diff
- Or you can use `make test-integration` which will implicitly install it for you.
+ Or you can use \`make test-integration\`, which will implicitly install it for you.
```

**Review by seshachalam-yv — COMMENTED** [review]
Overall, this looks good to me. Just a small nitpick:

Thanks, @unmarshall!

/lgtm

File: hack/prepare-envtest.sh (hunk 11 above)
Comment (seshachalam-yv) [reply hack/prepare-envtest.sh:19]: Alternatively, you could remove this check, as `prepare-env.sh` is invoked by `start-env.sh`, which in turn is triggered by the `make setup-envtest` target. This ensures that the `setup-envtest` command will be installed automatically when running the `make setup-envtest` target.

**Review by seshachalam-yv — APPROVED** [review]
